File: apps/mobile/lib/src/timeline/timeline_purge.dart
// SPEC-REF:
//   🔴 docs/rebuild/16-PORTABLE-RECORD-FORMAT-FPR-V1.md §6.2 (clear — five hard constraints)
//   docs/rebuild/15-DELIVERY-CHANNELS-STATES-AND-FAILURES.md G-17 / RV-96
//     (phone-side storage only ever grows, never shrinks — owner wants a
//     clear function he can see and operate)
//   docs/strategy/2026-08-01-data-asset-lifecycle-design.md §4-4, §5-4/-5/-6
//
// The clear feature's SELECTOR — a pure function, no I/O.
//
// ── Why the selector and the deleter are separate (the same shape on both ends) ───────────────────────────
// The confirmation dialog asks 「what will this delete」, the deleter does
// 「delete these」. **It must be the SAME batch of rows**,
// otherwise Book 16 §6.2-5's 「will free X」 is a guessed number, and a fake
// freed-space number and a fake progress bar
// are the same kind of thing. Once split into two halves, the UI gets exactly
// the list the deleter is about to consume next.
//
// ⚠️ The phone and the desktop **do not have the same number of delete
// triggers**, and that is not an inconsistency, it is a fact:
//   the desktop has three (capacity trimming / the user deleting a single row /
//   clear), the phone has only two —
//   **the phone's timeline has no automatic trimming at all** (SQLite keeps
//   everything, confirmed by Book 16 §4.1).
//   ⇒ On the phone 「clear」 is not 「a second delete path」, it IS **the
//   first**; this round folds in the existing
//   `TimelineStore.delete` too, so the two do not each delete half the job
//   (Book 15 G-21).

import 'timeline_entry.dart';

/// Which category to clear (owner's 2026-08-01 §4-4: 「clear transcribed text /
/// clear pictures, either can be run separately」).
enum ClearKind {
  /// Transcribed text rows.
  text,

  /// Picture rows (deletes the local original file along with it — that is
  /// the space owner wants freed).
  images,

  /// Both.
  both,
}

/// owner's five time windows, plus a sixth that **he did NOT rule on**, marked
/// as such everywhere it appears.
///
/// 🔴 [all] is the lead's **assumption**, pending owner's B5 approval
/// (`docs/decisions/2026-08-02-b5-stats-list-and-clear-all-options.md` §2),
/// **not a ruling**. owner listed only five windows, and 「a week ago」 does
/// not include the most recent week ⇒ per the windows he listed
/// **there is today no way at all to clear the timeline completely**, and that
/// is exactly what a device-switch/privacy scenario needs.
/// Deleting it only requires deleting this one enum value: the UI is generated
/// from this list.
enum ClearWindow { week, month, quarter, halfYear, year, all }

/// Days per window. **Fixed day counts, no calendar arithmetic**: 「three
/// months ago」 must be the same
/// span in February and in July — a deletion whose range the user cannot
/// predict is far worse than one that is off by a few days from intuition.
const Map<ClearWindow, int> kClearWindowDays = <ClearWindow, int>{
  ClearWindow.week: 7,
  ClearWindow.month: 30,
  ClearWindow.quarter: 90,
  ClearWindow.halfYear: 182,
  ClearWindow.year: 365,
};

/// The instant this window points to: rows **strictly earlier than it** are in range.
///
/// [ClearWindow.all] returns `null` — 「no lower bound」 and 「the lower bound is
/// now」 are two different things,
/// so this returns null rather than `now`.
DateTime? horizonOf(ClearWindow window, DateTime now) {
  final int? days = kClearWindowDays[window];
  if (days == null) return null;
  return now.subtract(Duration(days: days));
}

/// ⚠️ REQ-12-13 explicit note: **a remote-key-press row falls into the "text"
/// category** (it is not a picture, and [ClearKind]
/// only has two selectable types). This is a **deliberate choice this round,
/// not an oversight**:
///   · adding a third category would require touching [ClearKind], the
///     four-language copy, and the selector strip generated off
///     `ClearKind.values`
///     — that is a **product-level** change (the user has one more concept to
///     learn), belongs to owner, not to this cut;
///   · whereas 「clear transcribed text (last 7 days)」 clears the same-period
///     key-press rows along with it, consistent with the user's intent:
///     what they are clearing is the record from that period.
/// ⇒ Registered in Book 15 §2.0-e's residuals; the day owner wants them
/// cleared separately, adding a third category will fail to compile
/// (the switch in `stats_clear_sheet.dart` is exhaustive).
bool _inKind(TimelineEntry e, ClearKind kind) {
  if (kind == ClearKind.both) return true;
  return kind == ClearKind.images ? e.isImage : !e.isImage;
}

/// Which rows one clear operation will delete. **Pure**: the confirmation box
/// counts it, the deleter deletes it, the two cannot disagree.
///
/// Already-soft-deleted rows are not counted again — to the user it is already
/// gone, and counting it into 「will delete N rows」 would make that number
/// impossible to justify.
List<TimelineEntry> planClear(
  Iterable<TimelineEntry> all,
  ClearKind kind,
  DateTime? olderThan,
) {
  return all
      .where((TimelineEntry e) {
        if (e.deleted || !_inKind(e, kind)) return false;
        if (olderThan == null) return true;
        return e.createdAt.isBefore(olderThan);
      })
      .toList(growable: false);
}

/// 「everything before this instant is gone」 — **one per category**.
///
/// 🔴 This is the structure for 「cleared ≠ never existed」 (unified design
/// §5-5) under the constraint of **not over-claiming**.
/// If only pictures were cleared but a single unified cutoff were advanced,
/// the interface would say 「records before X have all been cleared」,
/// while every text row from that period is still on screen — a single scalar
/// does not even have the capacity to express this.
///
/// ⚠️ **The phone previously had no cutoff at all** (confirmed by Book 16
/// §4.1: SQLite keeps everything, no trimming whatsoever),
/// so this is not 「porting over the desktop's version」, it is **the first
/// time the phone can answer "what have I cleared"**.
class Cutoffs {
  const Cutoffs({this.text, this.images});

  static const Cutoffs none = Cutoffs();

  /// The instant of the most recent **text** row this device has deleted; null
  /// = nothing has ever been deleted.
  final DateTime? text;

  /// Same, for **picture** rows.
  final DateTime? images;

  /// The sentence the interface keeps saying: 「everything before X has been
  /// cleared」.
  ///
  /// = the **earlier** of the two marks, and an answer is only given when
  /// **both** are present: if pictures were cleared up to March but
  /// not a single text row has ever been cleared, there is no instant for
  /// which 「nothing before it exists」 holds.
  /// Deliberately says less rather than more — the interface falls back to the
  /// two per-category sentences, and those two are true.
  DateTime? get combined {
    final DateTime? t = text;
    final DateTime? i = images;
    if (t == null || i == null) return null;
    return t.isBefore(i) ? t : i;
  }

  bool get isEmpty => text == null && images == null;

  Map<String, Object?> toJson() => <String, Object?>{
    'text': text?.toIso8601String(),
    'images': images?.toIso8601String(),
  };

  static Cutoffs fromJson(Object? raw) {
    if (raw is! Map) return Cutoffs.none;
    DateTime? at(Object? v) =>
        v is String && v.isNotEmpty ? DateTime.tryParse(v)?.toUtc() : null;
    return Cutoffs(text: at(raw['text']), images: at(raw['images']));
  }
}

/// Advances the marks based on **what was actually deleted**.
///
/// Only forward (a later deletion never undoes an earlier one), and only
/// advances the categories this deletion actually covered.
/// `advance == null` is a **single-row deletion**: the user deleted one row,
/// which says nothing about 「everything before it is gone」,
/// so the marks do not move — that is exactly the distinction between 「I
/// deleted this one」 and 「everything before it is gone」.
///
/// 🔴 [ClearKind.both] advances **both** marks using **the most recent instant
/// among all the deleted rows**, rather than taking
/// each category's own most recent: a deletion that does not distinguish by
/// category deletes 「everything earlier than its most recent row, regardless of type」.
Cutoffs advanceCutoffs(
  Cutoffs current,
  Iterable<TimelineEntry> dropped,
  ClearKind? advance,
) {
  if (advance == null || dropped.isEmpty) return current;
  DateTime? newest(ClearKind kind) {
    DateTime? max;
    for (final TimelineEntry e in dropped) {
      if (!_inKind(e, kind)) continue;
      if (max == null || e.createdAt.isAfter(max)) max = e.createdAt;
    }
    return max;
  }

  DateTime? bump(DateTime? cur, DateTime? mark) {
    if (mark == null) return cur;
    if (cur == null || mark.isAfter(cur)) return mark;
    return cur;
  }

  if (advance == ClearKind.both) {
    final DateTime? mark = newest(ClearKind.both);
    return Cutoffs(
      text: bump(current.text, mark),
      images: bump(current.images, mark),
    );
  }
  return Cutoffs(
    text: advance == ClearKind.images
        ? current.text
        : bump(current.text, newest(ClearKind.text)),
    images: advance == ClearKind.text
        ? current.images
        : bump(current.images, newest(ClearKind.images)),
  );
}

File: apps/desktop/src/lib/timeline-purge.ts
// SPEC-REF:
//   🔴 docs/rebuild/16-PORTABLE-RECORD-FORMAT-FPR-V1.md §6.2 (clearing — five hard constraints)
//   docs/rebuild/15-DELIVERY-CHANNELS-STATES-AND-FAILURES.md G-21 (a single-row delete
//     must not delete bytes), G-17 / RV-96 (storage grows monotonically; this module is
//     where that gets closed off)
//   docs/strategy/2026-08-01-data-asset-lifecycle-design.md §5-4/-5/-6
//
// 🔴 THE ONE DELETER. Every row that leaves this PC leaves through [purge].
//
// ── WHY A MODULE AND NOT THREE CALL SITES ───────────────────────────────────
// The umbrella design doc §5-6, verbatim: "user-triggered clear and automatic trimming
// are two triggers for the SAME delete implementation; there must not be two separate
// deletion code paths (otherwise 'is it really all gone' has two different answers)".
// That is not a style preference — the repo had already paid for the second path
// before this file existed:
//
//   automatic trimming (TimelineStore.evict)  dropped the row AND its picture file ✅
//   user single-row delete (TimelineStore.remove) dropped the row and LEFT the picture ❌  ← G-21
//
// One question ("are this row's bytes gone?"), two answers, and the wrong one was the
// one the user reaches by tapping Delete. So the three TRIGGERS stay separate and the
// DELETER is this function. A fourth trigger tomorrow inherits the fix by construction.
//
// ── WHAT IT DELETES ─────────────────────────────────────────────────────────
// A desktop row owns two things: itself, and — for a picture row that kept its
// delivered bytes (`full_image`) — a FILE beside the forensic log. Both go. The
// remembered image-id chip goes with them, because a chip for a row that no longer
// exists is a claim about nothing.
//
// ⚠️ The phone additionally owns a third thing (the unknown-field vault, book 16 §9b-4);
// the desktop has no such side table — its carried fields ride on the row itself
// (lib/portable/preserved.ts) and therefore die with it. Stated because the two ends'
// deleters look asymmetric and the asymmetry is real, not an omission.

import type { TimelineRow } from './types';

/** WHICH KIND of row a clear is about (owner 2026-08-01 §4-4: "clear transcript text /
 *  clear images — the two can be exercised independently"). `both` is what automatic
 *  trimming does — it is about capacity, not kind. */
export type ClearKind = 'text' | 'images' | 'both';

/** owner's five time buckets (§4-4), plus the sixth that is NOT his and is marked as
 *  such everywhere it appears.
 *
 *  🔴 `all` is a LEAD CONTROLLER ASSUMPTION pending owner's B5 answer
 *  (`docs/decisions/2026-08-02-b5-stats-list-and-clear-all-options.md` §2), NOT a
 *  ruling. owner listed five buckets and "older than a week" does not include the last
 *  week, so as specified there is no way to empty the timeline at all — which the
 *  device-switch / privacy scenarios need. It ships behind the same confirmation as
 *  the rest and is trivially removable (delete the entry; `horizonOf` and the UI are
 *  data-driven off this list). */
export const CLEAR_WINDOWS = ['week', 'month', 'quarter', 'halfYear', 'year', 'all'] as const;
export type ClearWindow = (typeof CLEAR_WINDOWS)[number];

/** Days per bucket. FIXED DAY COUNTS, not calendar arithmetic, on purpose: "older than
 *  three months" has to mean the same span in February as in July, and a deletion the user cannot
 *  predict the extent of is worse than one that is a few days off their intuition. */
const WINDOW_DAYS: Record<Exclude<ClearWindow, 'all'>, number> = {
  week: 7,
  month: 30,
  quarter: 90,
  halfYear: 182,
  year: 365,
};

/** The instant a bucket points at: rows STRICTLY older than it are in scope.
 *  `null` for `all` — "there is no lower bound", which is a different thing from "the
 *  lower bound is now" and is why this returns null rather than
 *  `new Date(now).toISOString()`. */
export function horizonOf(window: ClearWindow, nowMs: number): string | null {
  if (window === 'all') return null;
  return new Date(nowMs - WINDOW_DAYS[window] * 86_400_000).toISOString();
}

/** Whether a row is in scope for `kind`.
 *
 *  A picture row is "an image" whatever its text says, and a transcript row is "text" —
 *  the split is `entry_type`, the same field the export's two-way choice and the
 *  timeline's chip read, so a row cannot be one kind here and another kind there. */
function inKind(row: TimelineRow, kind: ClearKind): boolean {
  if (kind === 'both') return true;
  return kind === 'images' ? row.entry_type === 'image' : row.entry_type !== 'image';
}

/** THE SELECTOR for the user's trigger: which rows a clear would remove.
 *
 *  Pure and separate from [purge] so the confirmation dialog can ask "what will this
 *  delete" and get the SAME answer the deletion will act on — book 16 §6.2-5: "must
 *  not multiply a row count by an average". The dialog counts these rows; the deleter
 *  deletes these rows.
 *
 *  A row with no usable `created_at` (`''`, possible on a row normalised from an
 *  ancient cache) is NEVER selected by a time-bounded clear: it would compare below
 *  every real instant and be swept up by "older than a year" while genuinely being of
 *  unknown age. It IS selected by `all`, which makes no claim about age. */
export function planClear(
  all: readonly TimelineRow[],
  kind: ClearKind,
  olderThan: string | null,
): TimelineRow[] {
  return all.filter((r) => {
    if (!inKind(r, kind)) return false;
    if (olderThan === null) return true;
    return r.created_at !== '' && r.created_at < olderThan;
  });
}

/** "Everything before this instant is gone" — per kind, because a type-filtered clear
 *  cannot honestly claim anything about the kind it did not touch.
 *
 *  🔴 THIS IS THE STRUCTURE THAT KEEPS "cleared ≠ never existed" TRUE (umbrella design
 *  doc §5-5) WITHOUT OVER-CLAIMING. One scalar cutoff was right while the only deleter
 *  was capacity eviction (which drops rows regardless of kind). Clearing only pictures
 *  and then advancing that one scalar would tell the user "records before X have all
 *  been cleared" while every transcript from that period is still on screen — a lie
 *  the old shape could not even express. */
export interface Cutoffs {
  /** Newest transcript instant this PC has dropped, ever. `null` = never dropped one. */
  text: string | null;
  /** Newest picture instant this PC has dropped, ever. */
  images: string | null;
}

export const NO_CUTOFFS: Cutoffs = { text: null, images: null };

/** The one sentence the page has always shown: "everything (any record) before X has
 *  been cleared".
 *
 *  = the OLDER of the two per-kind marks, and `null` unless BOTH are known: if
 *  pictures have been dropped back to March but no transcript ever has, there is no
 *  instant before which everything is gone. Under-claiming on purpose — the page
 *  falls back to the per-kind lines, which are true. */
export function combinedCutoff(c: Cutoffs): string | null {
  if (c.text === null || c.images === null) return null;
  return c.text < c.images ? c.text : c.images;
}

/** Read the persisted shape, tolerating the pre-C2 one.
 *
 *  0.2.37 and earlier wrote `{cutoff: "<iso>"}` — a single scalar written ONLY by
 *  capacity eviction, which drops both kinds. So an old value maps to BOTH marks, and
 *  that is a faithful translation rather than a guess: at the moment it was written,
 *  every kind older than it really had been dropped. */
export function readCutoffs(raw: unknown): Cutoffs {
  const o = raw as { cutoff?: unknown; text?: unknown; images?: unknown } | null;
  const str = (v: unknown): string | null => (typeof v === 'string' && v !== '' ? v : null);
  const legacy = str(o?.cutoff);
  if (legacy !== null && o?.text === undefined && o?.images === undefined) {
    return { text: legacy, images: legacy };
  }
  return { text: str(o?.text), images: str(o?.images) };
}

/** Push the marks forward by what was actually dropped.
 *
 *  Only ever forward (a later deletion cannot un-delete an earlier one), and only for
 *  kinds this deletion really covered. `advance === null` is a SINGLE-ROW delete: the
 *  user removing one line says nothing about everything older than it, so the marks do
 *  not move — that is the difference between "I deleted this one" and "everything
 *  before it is gone". */
export function advanceCutoffs(
  current: Cutoffs,
  dropped: readonly TimelineRow[],
  advance: ClearKind | null,
): Cutoffs {
  if (advance === null || dropped.length === 0) return current;
  // 🔴 `both` uses the newest instant across ALL dropped rows for BOTH marks, NOT each
  // kind's own newest — and that asymmetry with the filtered case is the whole point.
  // A kind-blind deletion (capacity eviction, or an "all" clear) removes everything
  // older than its newest casualty WHATEVER KIND it was, so both marks are earned even
  // when the batch happened to contain only transcripts. Taking each kind's own newest
  // here would leave `images` null on a transcript-only eviction and silently retire the
  // "everything before X has been cleared" line that has shipped since 0.2.26.
  const newest = (kind: ClearKind): string | null => {
    let max: string | null = null;
    for (const r of dropped) {
      if (r.created_at === '' || !inKind(r, kind)) continue;
      if (max === null || r.created_at > max) max = r.created_at;
    }
    return max;
  };
  const bump = (cur: string | null, mark: string | null): string | null =>
    mark !== null && (cur === null || mark > cur) ? mark : cur;
  if (advance === 'both') {
    const mark = newest('both');
    return { text: bump(current.text, mark), images: bump(current.images, mark) };
  }
  return {
    text: advance === 'images' ? current.text : bump(current.text, newest('text')),
    images: advance === 'text' ? current.images : bump(current.images, newest('images')),
  };
}

/** What [purge] mutates. A narrow view of the store rather than the store itself, so
 *  this module never learns about persistence, listeners or search — and so a test can
 *  hand it two plain collections. */
export interface PurgeSink {
  /** Keyed by `rowKey(channel, id)`, exactly as TimelineStore holds them. */
  rows: Map<string, TimelineRow>;
  /** The remembered image-id chips (R6 T-4). */
  imageIds: Set<string>;
  /** Delete these rows' picture FILES. Ids are bare row ids — both channels share one
   *  picture directory (`row_image.rs`). */
  dropPictures(ids: readonly string[]): void;
  /** `rowKey`, injected so this module does not import the address rules it does not
   *  own. */
  keyOf(row: TimelineRow): string;
}

/** What one deletion did — for the caller to report, never to guess at. */
export interface PurgeResult {
  /** Rows actually removed (a row named twice, or already gone, counts once). */
  removed: number;
  /** Picture FILES this deletion asked the filesystem to drop. ≤ [removed], and the
   *  gap is real: a picture row whose bytes were never kept has none. */
  picturesDropped: number;
  /** The marks after this deletion. */
  cutoffs: Cutoffs;
}

/** 🔴 THE DELETER. Nothing else in the desktop removes a timeline row.
 *
 *  It is deliberately NOT responsible for persisting or notifying: the callers already
 *  own a `persist()`/`notify()` pair with their own ordering rules (eviction runs
 *  INSIDE persist, so a persist here would recurse). What it IS responsible for is
 *  that every byte belonging to a removed row goes with it. */
export function purge(
  sink: PurgeSink,
  doomed: readonly TimelineRow[],
  advance: ClearKind | null,
  cutoffs: Cutoffs,
): PurgeResult {
  if (doomed.length === 0) return { removed: 0, picturesDropped: 0, cutoffs };
  const gone: TimelineRow[] = [];
  for (const r of doomed) {
    const key = sink.keyOf(r);
    if (!sink.rows.has(key)) continue;
    sink.rows.delete(key);
    gone.push(r);
  }
  if (gone.length === 0) return { removed: 0, picturesDropped: 0, cutoffs };
  // The chip goes with the row it describes — for EVERY removed row, not only picture
  // ones: the set is keyed by bare id and is what makes a row render as an image, so a
  // leftover entry would re-label a future row that happens to reuse the id.
  for (const r of gone) sink.imageIds.delete(r.id);
  // 🔴 G-21's fix. `full_image` is the WRITE's own verdict that bytes were kept, so
  // this asks the filesystem only about files that exist — and asks for EVERY removal
  // path, which is the whole point of there being one deleter.
  const pics = [...new Set(gone.filter((r) => r.full_image).map((r) => r.id))];
  if (pics.length > 0) sink.dropPictures(pics);
  return {
    removed: gone.length,
    picturesDropped: pics.length,
    cutoffs: advanceCutoffs(cutoffs, gone, advance),
  };
}
